### PR TITLE
Implement prefs and other functionalities for Tectonic trees

### LINKED ISSUE #5333: Add tree prefs for Tectonic trees and verify Tectonic tree functionalities
After #5318 and #5316

- [ ] Add a count stat to Tectonic trees to make `TreeEditor.Rank.Threshold.TectonicUnit` pref work (info: https://github.com/specify/specify7/pull/5318#discussion_r1799887113)
- [ ] Verify `queryFromTree` works and fix unit tests for those (info: https://github.com/specify/specify7/pull/5318#discussion_r1797332831)
- [ ] Add static data for Tectonic trees and verify tree unit tests work

### PR DESCRIPTION
**Pull request by sharadsw**
Fixes #5333

- [x] Add a count stat to Tectonic trees to make `TreeEditor.Rank.Threshold.TectonicUnit` pref work (info: https://github.com/specify/specify7/pull/5318#discussion_r1799887113)
- [x] Verify `queryFromTree` works and fix unit tests for those (info: https://github.com/specify/specify7/pull/5318#discussion_r1797332831)
- [x] Add static data for Tectonic trees and verify tree unit tests work

- Added workbench scoping for Tectonic trees
- Added tree hierarchy context for Tectonic trees

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [x] Add automated tests
- [x] Add relevant issue to release milestone

### Testing instructions

- **Use a paleo db**
- Go to Trees > Tectonic Unit
  - [ ] Verify a node can be added
  - [ ] Verify a node can be removed
  - [ ] Verify a rank can be added
  - [ ] Verify a rank can be removed
  - [ ] Tree can be renamed
  - [ ] Verify a node can be searched
  - Test other tree functionalities if any
 - Go to Query Builder
   - [ ] Verify queries work for Tectonic Unit 
   - [ ] Verify queries work for Tectonic Unit Tree Definition
   - [ ] Verify queries work for Tectonic Unit Tree Definition Items
- Go to workbench
  - [ ] Verify Tectonic Unit can be added via Workbench

### CODE REVIEW

File: specifyweb/frontend/js_src/lib/components/Molecules/SvgIcon.tsx [hunk 1]
```diff
@@ -239,6 +239,9 @@ const colorMapper = f.store<Partial<RR<keyof Tables, Gradient>>>(() => ({
   ...startsWith('DNA', colors.purple),
   ...startsWith('Sp', colors.lightBlue),
   ...startsWith('Workbench', colors.green),
+  ...startsWith('AbsoluteAge', colors.blue),
+  ...startsWith('RelativeAge', colors.lightBlue),
+  ...startsWith('TectonicUnit', colors.red),
```
Comment (sharadsw) [inline specifyweb/frontend/js_src/lib/components/Molecules/SvgIcon.tsx:244]: Going with these icon colors for the age tables for now to fix the unit test for this. Let me know if something else would be preferable

**Review by CarolineDenis — APPROVED** [review]

**Comment by emenslin** [comment]
Testing instructions?

**Comment by sharadsw** [comment]
@combs-a Thanks! It looks like that db either had a faulty migration or the migration had already ran once before the migration file name was changed and so it doesn't have the latest changes. I tried manually re-applying the migration but that doesn't seem to work right now 
```
  Applying specify.0008_tectonic_ranks...Traceback (most recent call last):
  File "manage.py", line 25, in <module>
    execute_from_command_line(sys.argv)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/core/management/base.py", line 89, in wrapped
    res = handle_func(*args, **kwargs)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 244, in handle
    post_migrate_state = executor.migrate(
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/db/migrations/executor.py", line 227, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/db/migrations/migration.py", line 123, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/db/migrations/operations/special.py", line 190, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/opt/specify7/specifyweb/specify/migrations/0008_tectonic_ranks.py", line 130, in consolidated_python_django_migration_operations
    create_root_tectonic_node(apps)
  File "/opt/specify7/specifyweb/specify/migrations/0008_tectonic_ranks.py", line 87, in create_root_tectonic_node
    tectonic_tree_def, created = TectonicUnitTreeDef.objects.get_or_create(
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/db/models/query.py", line 581, in get_or_create
    return self.get(**kwargs), False
  File "/opt/specify7/ve/lib/python3.8/site-packages/django/db/models/query.py", line 439, in get
    raise self.model.MultipleObjectsReturned(
__fake__.MultipleObjectsReturned: get() returned more than one TectonicUnitTreeDef -- it returned 2!
```
Can you try with a different paleo db in the mean time?

**Review by emenslin — CHANGES_REQUESTED** [review]
### Testing instructions



- **Use a paleo db**
- Go to Trees > Tectonic Unit
  - [x] Verify a node can be added
  - [x] Verify a node can be removed
  - [x] Verify a rank can be added
  - [ ] Verify a rank can be removed
  - [x] Tree can be renamed
  - [x] Verify a node can be searched
  - Test other tree functionalities if any
 - Go to Query Builder
   - [x] Verify queries work for Tectonic Unit 
   - [x] Verify queries work for Tectonic Unit Tree Definition
   - [x] Verify queries work for Tectonic Unit Tree Definition Items
- Go to workbench
  - [ ] Verify Tectonic Unit can be added via Workbench

--- 
I did run into the same problems as https://github.com/specify/specify7/pull/5345#pullrequestreview-2402824855 but most I could solve by creating new forms

- Cannot delete any ranks that were made with the autogenerated forms, it creates a deletion blocker loop. If you create new forms and then add a rank those can be deleted. 


https://github.com/user-attachments/assets/764e8d7d-64f7-4556-9ad0-936b9986cdc3




- If you try to add tectonic unit in the workbench it throws an error 

https://github.com/user-attachments/assets/fe8023d8-f091-4a36-b6fe-9e02cb6d19ef


[Specify 7 Crash Report - 2024-10-29T20_45_34.891Z.txt](https://github.com/user-attachments/files/17562953/Specify.7.Crash.Report.-.2024-10-29T20_45_34.891Z.txt)

**Comment by sharadsw** [comment]
Added the migration tag to this PR. Although this PR doesn't directly have migration changes, the PR it is based on does due to which this PR needs to be tested in the same way

**Comment by sharadsw** [comment]
@emenslin @combs-a Can this be re-tested following the new steps for testing migration related PRs? Unfortunately it looks like both `ChadronTest` and `KUVertPaleo` have migration issues

**Review by pashiav — CHANGES_REQUESTED** [review]
### Testing instructions



- **Use a paleo db**
- Go to Trees > Tectonic Unit
  - [X] Verify a node can be added
  - [X] Verify a node can be removed
  - [X] Verify a rank can be added
  - [ ] Verify a rank can be removed
  - [X] Tree can be renamed
  - [X] Verify a node can be searched
  - Test other tree functionalities if any
 - Go to Query Builder
   - [X] Verify queries work for Tectonic Unit 
   - [X] Verify queries work for Tectonic Unit Tree Definition
   - [X] Verify queries work for Tectonic Unit Tree Definition Items
- Go to workbench
  - [X] Verify Tectonic Unit can be added via Workbench


---
  - [ ] Verify a rank can be removed
		- https://github.com/specify/specify7/issues/5025
		- Ran into the same issue as https://github.com/specify/specify7/pull/5345#pullrequestreview-2403054066
		

> Cannot delete any ranks that were made with the autogenerated forms, it creates a deletion blocker loop. If you create new forms and then add a rank those can be deleted.

- Root's full name is blank when queried. This behavior is inconsistent.
Screen Shot 2024-10-30 at 1 57 04 PM
Screen Shot 2024-10-30 at 1 56 45 PM

<br>

- Viewing Root on the Data Entry form, the Clone and Add buttons seem to be blocked, but you can still click on them and see they are functional: https://oldchadrontest2migration-issue-5333.test.specifysystems.org/specify/view/tectonicunit/1/


https://github.com/user-attachments/assets/d068b6bd-c0f5-4d7d-ab85-4bbf9febbb41


<br>

- For the autogenerated form, when you type in "Tectonic Unit" into the **Definition** QCB, it doesn't pop up as an option, even though it is the only option. **Definition Item** and **Parent** QCBs' options also do not show up no matter what the input.

https://github.com/user-attachments/assets/81782443-32b4-4cf1-a978-ea15d572ff8b

**Review by Areyes42 — APPROVED** [review]
### Testing instructions

-   **Use a paleo db**
-   Go to Trees > Tectonic Unit
    -  [x]  Verify a node can be added
    -   [x] Verify a node can be removed
    -   [x] Verify a rank can be added
    -   [ ] Verify a rank can be removed -see comment
    -   [x] Tree can be renamed
    -   [x] Verify a node can be searched
    -   Test other tree functionalities if any
-   Go to Query Builder
    -   [x] Verify queries work for Tectonic Unit
    -   [x] Verify queries work for Tectonic Unit Tree Definition
    -   [x] Verify queries work for Tectonic Unit Tree Definition Items
-   Go to workbench
    -   [x] Verify Tectonic Unit can be added via Workbench
***
Everything looks good after running the migrations! Ran into an issue (https://github.com/specify/specify7/issues/5025) when testing rank removals, but looks to be a separate issue. Rank removal still works for lowest tiered ranks that are unassociated.

**Comment by sharadsw** [comment]
@pashiav I'm not able to reproduce the blank full name locally or on `KUInvertPaleo` in the test panel. It looks like a db specific issue (hopefully not a migration issue). Also, I think the root node will probably be renamed by users later on anyway.

The QCBX search doesn't work because of missing type search definitions. I have added type search definitions for them now but it looks like the changes from the static file are not detected. If that's the case, make sure to add these lines to the Type Searches app resource. 
```XML
<typesearch tableid="1025"  name="TectonicUnitTreeDef"       searchfield="name"            displaycols="name"               format="%s"     dataobjformatter=""/>
<typesearch tableid="1026"  name="TectonicUnitTreeDefItem"       searchfield="name"            displaycols="name"               format="%s"     dataobjformatter=""/>
<typesearch tableid="1027"  name="TectonicUnit"       searchfield="name"            displaycols="name"               format="%s"     dataobjformatter=""/>
```

I notice other tree forms don't have these QCBX. I think when we add a default viewdef for tectonic trees, those won't be needed and also clone/add won't be blocked in data entry as a result. It's being blocked due to the empty qcbxs in the form

**Review by emenslin — APPROVED** [review]
### Testing instructions



- **Use a paleo db**
- Go to Trees > Tectonic Unit
  - [x] Verify a node can be added
  - [x] Verify a node can be removed
  - [x] Verify a rank can be added
  - [x] Verify a rank can be removed
  - [x] Tree can be renamed
  - [x] Verify a node can be searched
  - Test other tree functionalities if any
 - Go to Query Builder
   - [x] Verify queries work for Tectonic Unit 
   - [x] Verify queries work for Tectonic Unit Tree Definition
   - [x] Verify queries work for Tectonic Unit Tree Definition Items
- Go to workbench
  - [x] Verify Tectonic Unit can be added via Workbench

Looks good!

**Review by combs-a — DISMISSED** [review]
### Testing instructions

- **Use a paleo db**
- Go to Trees > Tectonic Unit
  - [X] Verify a node can be added
  - [X] Verify a node can be removed
  - [X] Verify a rank can be added
  - [ ] Verify a rank can be removed
  - [X] Tree can be renamed
  - [X] Verify a node can be searched
  - Test other tree functionalities if any
 - Go to Query Builder
   - [X] Verify queries work for Tectonic Unit 
   - [X] Verify queries work for Tectonic Unit Tree Definition
   - [X] Verify queries work for Tectonic Unit Tree Definition Items
- Go to workbench
  - [X] Verify Tectonic Unit can be added via Workbench

Did find a strange behavior with removing in that if you change the rank of a TreeDefItem, then you can no longer remove it from the tree, but not sure how to recreate it--I tested the same steps on the Taxon tree and tried to recreate it on the Tectonic Unit tree, but it worked as expected both times (still was able to delete the rank).

It says that it cannot be deleted because it's referenced by another tree rank, but it doesn't actually redirect to the tree rank when I click on it--it goes to the tree definition form.

image

Link to the tree with the undeleteable rank: https://chadrontest2migration5345-issue-5333.test.specifysystems.org/specify/tree/tectonicunit/?conformation=%7E-

**Comment by sharadsw** [comment]
@combs-a I'm not able to reliably recreate it either. Sounds like a strange bug, maybe it is related to trees that use rankIds in general and not specifically with Tectonic Unit? I've opened a new issue for it: #5364

**Review by acwhite211 — APPROVED** [review]

**Pull request merged**